### ISSUE TITLE
Roadmap: Nvim-treesitter 1.0

### ISSUE THREAD

**Issue by clason** [nvim-treesitter/nvim-treesitter#4767]
# Rationale

Nvim-treesitter was created to drive the user-space development of tree-sitter related features, including but not limited to highlighting. It basically offered three related but different features:

1. parser installation and management
2. query aggregation and creation
3. interface for other plugins to hook into as "modules"

This has been very successful, but the situation has changed significantly since it was started, with more and more parsers being added, while more actual features were included in Neovim core. At the same time, the current architecture is increasingly limiting further maintenance; in particular the module system is adding significant overhead and makes it hard to implement changes without breaking other plugins.

We should thus overhaul and rewrite nvim-treesitter to make it fit for the next few years. **This will be a breaking change for users and plugins**, but will hopefully allow us to mark the result as stable so the future experience will be much smoother. See below how this process will be handled.

# Goal

Nvim-treesitter should be leaner and focus on its core goals (1. and 2. above). Specifically,

## Parser installation

This plugin should **only handle installing, updating, and removing parsers and queries**:
 - parsers **and matching queries** should be installed in a standard directory (`stdpath('data')`, but configurable); ideally, you should not need to load the plugin if you only want to _use_ parsers;
 - in particular, only queries for installed parsers should be visible in `runtimepath`;
 - parsers should be sorted into **"tiers"**, e.g.,
   * _Tier 1_ parsers are stable (and ideally versioned), queries are maintained by the project itself, all tests must pass,
   * _Tier 2_ parsers are stable, queries are actively(!) maintained by the community, tests are run,
   * _Tier 3_ parsers are unstable, unmaintained, or marked experimental for other reasons.
   
   These tiers determine whether CI is run, whether parsers are automatically updated, and can be selected for installation (i.e., `ensure_installed = 'tier1'). Parsers should be aggressively moved between tiers as quality or maintenance changes.
 - only a limited set of reasonably popular platforms should be supported out-of-the-box for compilation; all others should be handled by making compilation commands configurable.
   

A big part of this will be pushing parser maintainers to use versioned releases (which may be a requirement for inclusion in Tier 1 or even Tier 2); one possible scheme would be to require
* _point releases_ for parser changes that do not affect queries at all (`grammar.js` or `scanner.c` bugfixes),
* _minor releases_ which _add_ nodes (meaning old queries will still work, but need changes to make full use of the updated grammar),
* _major releases_ which _remove_ or _rename_ nodes (meaning old queries will no longer work).

## Queries

We should move towards closer alignment with upstream and other editors using tree-sitter; this will make it easier to share efforts (and quicker to add new parsers or pull improvements from, e.g., Helix), see https://github.com/neovim/neovim/issues/22495

 - **injections**: should use the upstream format which is now supported by Neovim, see #4675 
 - **highlights**: where it makes sense, should use [Helix' captures](https://docs.helix-editor.com/themes.html#syntax-highlighting), which are derived from established SublimeText scopes. This merits a separate issue, but personally, I think we should at least switch from `@text.*` to
   * `@markup.*` and
   * `@comment.todo` etc.

We should also investigate how to streamline or automate syncing with upstream queries (if parser repos contain them).

It might also make sense to include more queries here such as `textobjects`, to make sure they track parser version closely. (Rule of thumb: every feature used -- or planned to be used -- in Neovim should have queries provided by nvim-treesitter.)

## Modules

We should completely drop the module framework; plugins should instead handle setup and attaching themselves.   (https://github.com/nvim-treesitter/nvim-treesitter-context demonstrates that this is quite simple.

In particular, the bundled modules should be removed:
* `indents` is a candidate for inclusion in Neovim; it's not quite ready for primetime yet, so it can stay for a while, but in the form of a simple `indentexpr` function instead of a module;
* `incremental-selection` mostly served as a proof-of-concept for non-highlighting uses of tree-sitter; if people are actively using it, they should consider moving it to a separate plugin (or seeing if `textobjects` don't serve this purposes even better);
* `textobjects` is also a candidate for inclusion; probably should be refactored into a plugin before that, though;
* `refactor` should be -- wait for it! -- refactored;
* `playground` is already mostly obsolete; the only missing feature (live query editor) is planned to be upstreamed.

## Plan

As this is a breaking change, we will have to work on this in parallel:

1. create a new `main` branch (targeting Neovim nightly _only_) for the new slimmed down installation functionality;
2. once that has a working prototype, move queries and make the breaking changes for upstream alignment;
3. at some point, move development to the new branch (only bugfixes on `master`; parser addition and query updates target  `main` and will be backported if desired);
4. after Neovim 0.10 is released, `main` will become the new default branch; the current `master` branch will remain (but frozen) for people requiring backwards compatibility;
5. make an 1.0 release targeting Neovim 0.10, and follow a proper release schedule (semver or otherwise) from then on
6. profit! 🥳

**Comment by theHamsta** [comment]
Having both `main` and `master` is probably confusing. Would suggest `develop` for the new thing until it becomes final.

**Comment by clason** [comment]
I don't think renaming branches is a great idea; there's some things that don't play nicely with it (e.g., Github Actions).

**Comment by lucario387** [comment]
Just adding this as a reminder for #4675: I haven't added back `include-children` to most `@injection.content` areas. 
This means it will differs from the 0.8-compat queries (and currently is breaking to some captures). But I'll wait for upstream neovim on what should the intended way to handle this be

**Comment by clason** [comment]
# Status (2025-12-14)

Steps 1-3 have been completed; Step 4 is where things fell apart... For various reasons, WASM parser support in Neovim will not happen in the near future; the adapted goal is to release 1.0 targeting Nvim 0.12 and tree-sitter 0.26.0 with the current functionality. **This means that the `main` branch can be considered "stable" in the sense that it is feature complete, from a user interface point of view.**

Here are the open issues across the ecosystem that need to be resolved before completing the transition with a 1.0 release:

### nvim-treesitter

While the functionality is fixed, there are some remaining tasks that possibly have user-facing consequences, so should be done before the release:

* The `async` module needs to be synced with the in-development `vim.async` module, which has undergone significant, breaking, changes since the version we vendored. 
* The list of parsers needs to be refreshed, aggressively moving parsers to Tier 1 if eligible or to Tier 3 if not (Tier 2 should be the exception, not the rule) and pruning inactive maintainers (or removing that information entirely; it has proved to be too ephemeral to be broadly useful).
* The `indents` module still [needs to be rewritten](https://github.com/nvim-treesitter/nvim-treesitter/issues/7840), but this can be done in a parallel feature branch to be merged for a 2.0 release.

**The default branch is now `main`**; at a later date, the `master` branch may be renamed to `legacy` (if enough plugin managers play nicely with that). **Note that the `master` branch is _not_ guaranteed to be compatible with Nvim 0.12 when it is released! To keep using `master`, you need to stick to Nvim 0.11.**

### nvim-treesitter-textobjects

This plugin now has a working compatible `main` branch, with the `master` branch frozen (but still default).

What remains to be done on `main`:

* adding convenience functions for setting up mappings.

### nvim-treesitter-refactor

This module was beyond saving; it is now archived in favor of a (more appropriately named) [nvim-treesitter-locals](https://github.com/nvim-treesitter/nvim-treesitter-locals), which bundles all locals-related functionality previously scattered over plugins. (No functionality yet; help is welcome, although much is blocked by the ongoing upstream `vim.pos` work.)

However, this is not a blocker for nvim-treesitter itself.
